reflow

drahnr opened the pull request:
<!--
Thank you for submitting a PR to cargo-spellcheck!
-->

## What does this PR accomplish?

Pushing #72 over the :checkered_flag: 

<!---
Delete all that do not apply:
-->

 * :peacock:  Feature

<!---
Mention the linked issue here.
This will magically close the issue once the PR is merged.
-->

Closes #39 

## Changes proposed by this PR:

<!---
Tell the reviewer What changed, Why, and How were you able to accomplish that?
-->

## Remaining Challenges

* [x] a span is not a 1:1 mapping of a line
* [ ] user selection must work with multiline suggestions (related to / covered by #95 )
* [ ] provide the same logic of `--code` exit code flags like we do for `check`
* [ ] assure test cases cover all variant types with all possible single/multiline variations

## Notes to reviewer:

<!---
Leave a message to whoever is going to review this PR.
Mainly, pointers to review the PR, and how they can test it.
If things are still WIP or feedback on particulr impl details
are wanted, state them here too.
-->


## 📜 Checklist

 * [ ] Works on the `./demo` sub directory
 * [ ] Test coverage is excellent and passes
 * [ ] Documentation is thorough

KuabeM commented on src/reflow/mod.rs:
```diff
@@ -0,0 +1,286 @@
+    where
+        'a: 's,
+    {
+        let wrapper = textwrap::Wrapper::new(config.max_line_length)
```
That's my old stuff right? Since it's never gonna work, maybe just remove it right away?

drahnr replied:
That would be the next step :+1:

drahnr commented on src/reflow/config.rs:
```diff
@@ -0,0 +1,17 @@
+impl Default for ReflowConfig {
+    fn default() -> Self {
+        Self {
+            max_line_length: 70,
```
Nope, but I think that was your code :smile_cat:

drahnr commented on src/reflow/iter.rs:
```diff
@@ -0,0 +1,294 @@
+    /// MUST be sorted by `.start` value.
+    /// Ranges must be in characters!
+    unbreakable_ranges: Vec<Range>,
+    unbreakable_idx: usize,
```
`unbreakable_idx` is an index into `unbreakable_ranges`, I tried to use an iterator first, but that did not turn out very pretty and just complicated things even more.

drahnr commented on src/reflow/iter.rs:
```diff
@@ -0,0 +1,294 @@
+    // * lineno
+    // * content of that line
+    // * and ranges relative to the original string which are covered by content
+    type Item = (Range, Range, Cow<'s, str>);
```
one is in characters and one is in bytes - so the first is necessary for accounting the length, the second for indexing into the initial string `s[range]`

drahnr commented on src/reflow/iter.rs:
```diff
@@ -0,0 +1,294 @@
+                // are we within an unbreakable area
+                // iff so, just continue until we are out of the weeds
+                if unbreakable.contains(&char_idx) {
+                    continue 'outer;
```
It's actually a label, it's a bit like goto without the suicide potential and reaaally handy for nested loops.

KuabeM replied:
Yeah I have to keep that in mind!

drahnr commented on src/reflow/iter.rs:
```diff
@@ -0,0 +1,294 @@
+            let item = (
+                char_range.clone(),
+                byte_range.clone(),
+                std::borrow::Cow::Borrowed(dbg!(&self.s[byte_range])),
```
As always...

KuabeM commented on src/reflow/iter.rs:
```diff
@@ -85,60 +85,69 @@ impl<'s> Iterator for Tokeneer<'s> {
 }
 
-/// Re-glue all tokenized items
+/// Re-glue all tokenized items under the constrained of a maximum line width
```
doc is really useful:+1:

drahnr commented on src/documentation/chunk.rs:
```diff
@@ -213,6 +213,23 @@ impl CheckableChunk {
             .collect::<IndexMap<_, _>>()
     }
 
+    pub fn find_spans_inclusive(&self, range: Range) -> Vec<Span> {
```
Docs :question:

drahnr replied:
And why inclusive?

drahnr replied:
And if inclusive, we should probably use `RangeInclusive` (roughly, iirc)

KuabeM replied:
That was more a working title.. I'm really bad at naming things though!

Inclusive should indicate that the all spans at least partially contain range, so all spans together contain the range if that makes any sense. I'll document this and tried to come up with a better name

drahnr commented on src/reflow/mod.rs:
```diff
@@ -59,11 +76,25 @@ fn reflow<'s>(origin: ContentOrigin, chunk: &'s CheckableChunk) -> Vec<Suggestio
             };
+
+            let spans = chunk.find_spans_inclusive(range.clone());
+            let span = Span {
```
we can't really guarantee it's none empty, can we? If so, let's use `expect("Does not X. qed")`

drahnr commented on src/reflow/mod.rs:
```diff
@@ -117,7 +148,9 @@ fn reflow<'s>(origin: ContentOrigin, chunk: &'s CheckableChunk) -> Vec<Suggestio
                     Tag::Paragraph => {
                         // regular end of paragraph
-                        paragraph = store(cover.end, unbreakables.as_slice());
+                        // @todo the cover.end is exclusive and find_span() of CheckableChunk is inclusive, such that find_span()
```
If this is the case, that is a bug. A `Range` by definition should never be inclusive.

drahnr replied:
I changed the todo fmt from `@todo` to `TODO`.

KuabeM replied:
I thought that it's a bug when I used `find_span()` before I introduced `find_spans_inclusive()`, so there is no bug. Removed it.

drahnr commented on src/reflow/mod.rs:
```diff
@@ -206,15 +274,65 @@ blanket text without any meaning",
+        let suggestions = suggestion_set
+            .iter()
+            .next()
+            .expect("Must contain exactly one item");
```
At parity the phrasing is different, but I can't seem to find the coding manual right now (used to be public at least).

KuabeM replied:
Why do you use  `qed`? The `expect()` should demonstrate that this never happens. So if the `expect()` is fired, we've shown that this assumption is wrong? More like `quod erat non demonstrandum` :stuck_out_tongue: But anyway, I'll use that phrase :)

KuabeM replied:
I'm not sure how to use this style. For instance, on the result of `reflow()` I should do something like `.expect("Reflowing works. qed");`?

KuabeM replied:
I think I get it after looking at your examples...

drahnr reviewed (commented):
Looks pretty well, I have yet to test this, a few nits as per usual :small_airplane:

drahnr commented:
Is `cargo spellcheck reflow /path/to/file` already supposed to work?

drahnr commented:
The assumption that there is one span per line is incorrect though.

Counter example:

```rust
#[doc=r#"yada
yada
"#]
```

drahnr commented:
Added initial hooks for args and forwarding all of that, there are still a few :construction: areas which need resolving.

* [ ] a span is not a 1:1 mapping of a line
* [ ] the user selection is completely borked with multline suggestion replacements
* [ ] :beers:

KuabeM commented:
Regarding the Span to line mapping: We can't really do something about that, the only (hacky) workaround I can think of is to count the number of whitespaces on multiline spans. In other words, use `span.start` for the first line of the span and we have to count the whitespaces for all subsequent lines.

drahnr commented on src/documentation/chunk.rs:
````diff
@@ -225,7 +225,22 @@ impl CheckableChunk {
+    /// |-- s1 --|-- s2 --|-- s3 --|-- s4 --|
+    ///             |----- range -----|
+    /// ```
+    ///
````
:heart:

drahnr commented on src/documentation/chunk.rs:
````diff
@@ -225,7 +225,22 @@ impl CheckableChunk {
+    ///             |----- range -----|
+    /// ```
+    ///
+    pub fn find_enclosing_spans(&self, range: Range) -> Vec<Span> {
````
```suggestion
    pub fn find_enclosed_spans(&self, range: Range) -> Vec<Span> {
```

drahnr replied:
Could we also return the iterator instead of calling `collect()`?

drahnr commented:
Not sure I understand your last comment. Do you mean newlines I stead of spaces? Not sure how that will help us here?

I think for `doc=#r"` comments we  would be able to reflow, for `doc="` the newlines would need to be escaped but any indent after the first would either be identical or set to 0.

drahnr commented on src/documentation/chunk.rs:
```diff
@@ -227,12 +227,14 @@ impl CheckableChunk {
+            .filter_map(|(fragment_range, fragment_span)| {
                 // could possibly happen on empty documentation lines with `///`
-                fragment_range.len() > 0
+                if fragment_range.len() > 0 {
```
```suggestion
                if fragment_range.is_empty() {
```
is preferable, but I forget about that too all the time..

drahnr commented on src/reflow/mod.rs:
```diff
@@ -447,12 +445,86 @@ should be rewrapped."#;
+            r#" Some more text after the rule which represents a
+ paragraph"#,
+        ];
+
```
The reflow check here is correct, but we could never use this as a direct replacement, since it would cover additional lines.
We would need a injection suggestion with a source span length of 0 (so nothing is removed, only text is inserted at that position). If this is the case, the comment type (`///` in this case ) needs to be added as a injection for any additional lines inserted.

drahnr commented on src/reflow/mod.rs:
```diff
@@ -44,13 +42,18 @@ impl Checker for Reflow {
 fn reflow_inner<'s>(
```
Without comments nobody will remember what reference `range` and `unbreakable_ranges` have

drahnr commented on src/reflow/mod.rs:
```diff
@@ -44,13 +42,18 @@ impl Checker for Reflow {
+    let s_absolute = &s[range.clone()];
+    let unbreakables = unbreakable_ranges
+        .iter()
+        .map(|r| (r.start - range.clone().start)..(r.end - range.clone().start));
```
Avoid the clones, `usize` is `Copy`.

drahnr commented on src/reflow/mod.rs:
```diff
@@ -97,21 +100,21 @@ fn reflow<'s>(
 
             // debug_assert!(!spans.is_empty());
```
Avoid remanescant dead code :crocodile:

drahnr commented on src/reflow/mod.rs:
```diff
@@ -125,7 +128,7 @@ fn reflow<'s>(
                 chunk.as_str(),
                 range.clone(),
-                unbreakable_ranges.to_vec(),
+                unbreakable_ranges,
```
:+1:

drahnr commented on src/reflow/mod.rs:
```diff
@@ -446,66 +450,70 @@ should be rewrapped."#;
 
-               with a second part that is fine"#]
-    struct Fluffy {};"##;
+        With a second part that is fine"#
```
Is the indentation here desired?

KuabeM replied:
Nope, thanks :)

drahnr commented on src/reflow/mod.rs:
```diff
@@ -234,14 +239,13 @@ mod tests {
 
-            // TODO add tests with indentation and unbreakables or is this sufficiently covered by Gluon tests?
-            let range = 0..CONTENT.len();
+            let range = 0..chunk.as_str().len();
```
This is in bytes, but we want chars. :warning:  It will give the idea of functionality but breaks down as soon as you you introduce emoji chars in your tests.
I'd recommend a couple :cactus: :tophat: :thermometer: to make sure that basic mistake is avoided. Could have prevented #107 and a few which I just uncovered in the process of debugging that.

drahnr commented on src/reflow/mod.rs:
```diff
@@ -32,13 +32,21 @@ impl Checker for Reflow {
+/// `range` denotes the range of the paragraph of interest in the top-level `CheckableChunk`.
+/// `unbreakable_ranges` contains all ranges of words/sequences which must not be split during
+/// the reflow. They are relative to the top-level `CheckableChunk` similar to `range`. The indentation
+/// vec contains the indentation for each line in `s`.
```
:heart:

drahnr commented on src/documentation/literal.rs:
```diff
@@ -12,6 +12,8 @@ use std::fmt;
     /// `///`
     TripleSlash,
+    /// `//!`
+    DoubleSlashEM,
```
:+1:

drahnr commented on src/documentation/literal.rs:
```diff
@@ -61,6 +63,9 @@ impl std::cmp::PartialEq for TrimmedLiteral {
         if self.span != other.span {
             return false;
         }
+        if self.variant != other.variant {
```
:+1:

drahnr commented on src/reflow/mod.rs:
```diff
@@ -64,26 +64,40 @@ fn reflow_inner<'s>(
     };
+    let mut reflow_applied = false;
+    let mut lines = s.lines();
+    let mut acc: String = if let Some((_, content, _)) = gluon.next() {
```
Could you elaborate on this?

KuabeM replied:
I agree, the check if a reflow happend is not really intuitive. It compares each line in the original string `s` with the result of the reflow, and right while I'm writing this comment, I see a bug. The comparison needs to happen on the absolute string since only that is fed into the `gluon`. I'll test and fix this and also add some comments :)

KuabeM replied:
Some more context: now the lines in the input to the `gluon` is compared to its output. If nothing changed, there was no reflow. Does that make more sense?

On the doc comment above: right now, the chunk shown by you would be reflown to `#[ doc = r#"a b c"]` if the the `max_line_width > 19 + indentation`. Is this not desired? I added a test for this exact case.

KuabeM replied:
another point about doc comments: we should keep the indentation of the second and following lines:

```
#[ doc = "First line
               Second line blabla"]
// should stay like this:
#[ doc = "First line Second
               line blabla"]
```
right?

drahnr reviewed (commented):
Looks good, needs a few more tests for various doc comment styles, and some verification on `demo` runs. I don't think there is too much work left now!

KuabeM commented:
One major blocker remains. If the original comment spans over two lines, but they would be changed by the reflow, the `correct_lines()` in `Action::correction()` panics on `There should be no multiline strings as of today: Start and end are not on the same line`. I think that needs changes in the `correction` code right?

drahnr commented:
Yeah, there might be a few edge cases which are not covered just yet.

I think investing some time in making sure we have all variants covered here gives you better feedback faster.

KuabeM commented:
I added some more fixes for the indentation issue:

```rust
struct Some {
   /// Lone line which will be changed
   member: i8,
   #[doc = "Another long line which will be changed"]
   sec: i8,
};
// will become
struct Some {
   /// Lone line which will
   /// be changed
   member: i8,
   #[doc = "Another long line
            which will be changed"]
   sec: i8,
};
```

Do you agree with this behavior?

drahnr commented:
I think that's the right thing to do, I can't test anything right now :) 🌴

KuabeM commented on src/action/mod.rs:
```diff
@@ -218,7 +247,7 @@ impl Action {
     /// Purpose was to check, checking complete, so print the results.
     fn check(&self, suggestions_per_path: SuggestionSet, _config: &Config) -> Result<Finish> {
-        let mut count = 0usize;
+        let mut count = 0_usize;
```
That's a clippy suggestion, I wanted to correct them all but got distracted on the way. How about a PR tackling clippy? (after this one)

drahnr replied:
We should first pick a sane set of clippy rules which should be enforced/ignored. I am not a big fan of `_` in that position.

drahnr commented:
I adjusted the set of tests as I anticipated them. It would make sense to write down your target behaviour and write all test cases before tackling the implementation before, so this stops being a moving target for you - I added the simple `#[doc="x"]` variant test, note the spans are set on a best effort basis for now and should be double checked.

drahnr commented:
And we could possible make an bandaid possibly an enum, with 3 variants: `Replacement(Span, String)`, `Injection(LineColumn, String)`, `Deletion(Span)` - but this might be more of a refactoring in a secondary refactoring round.

KuabeM commented:
> I adjusted the set of tests as I anticipated them. It would make sense to write down your target behaviour and write all test cases before tackling the implementation before, so this stops being a moving target for you - I added the simple `#[doc="x"]` variant test, note the spans are set on a best effort basis for now and should be double checked as needed.
> 
> There is also a remaining open question that has yet to be addressed:
> 
> ```rust
> #[doc="a"]
> #[doc="b"]
> struct X;
> ```
> 
> should be optionally (!) transformed to
> 
> ```rust
> #[doc="a b"]
> struct X;
> ```
> 
> I think, open for pros and cons.
> 
> The opposite direction.
> 
> ```rust
> #[doc="aaaa bbbb"]
> struct X;
> ```
> 
> should be optionally (!) transformed to
> 
> ```rust
> #[doc="aaaa"]
> #[doc="bbbb"]
> struct X;
> ```
> 
> **OR**
> 
> ```rust
> #[doc="aaaa
>             bbbb"]
> struct X;
> ```
> 
> but for the start we should pick one and stick with it throughout, open for opinions as well :)

I'm not sure if I like such behaviour. If the original content had to distinct lines of doc, I'd assume that it was on purpose.

drahnr commented:
> > I adjusted the set of tests as I anticipated them. It would make sense to write down your target behaviour and write all test cases before tackling the implementation before, so this stops being a moving target for you - I added the simple `#[doc="x"]` variant test, note the spans are set on a best effort basis for now and should be double checked as needed.
> > There is also a remaining open question that has yet to be addressed:
> > ```rust
> > #[doc="a"]
> > #[doc="b"]
> > struct X;
> > ```
> > 
> > 
> > should be optionally (!) transformed to
> > ```rust
> > #[doc="a b"]
> > struct X;
> > ```
> > 
> > 
> > I think, open for pros and cons.
> > The opposite direction.
> > ```rust
> > #[doc="aaaa bbbb"]
> > struct X;
> > ```
> > 
> > 
> > should be optionally (!) transformed to
> > ```rust
> > #[doc="aaaa"]
> > #[doc="bbbb"]
> > struct X;
> > ```
> > 
> > 
> > **OR**
> > ```rust
> > #[doc="aaaa
> >             bbbb"]
> > struct X;
> > ```
> > 
> > 
> > but for the start we should pick one and stick with it throughout, open for opinions as well :)
> 
> I'm not sure if I like such behaviour. If the original content had to distinct lines of doc, I'd assume that it was on purpose.

The idea is to make it configurable. Note that i.e. bindgen would generate such messages, so it's not necessarily intended but just default behaviour of another tool which is just incidental, not intentional.

But Mostly the first case is relevant for now, but that needs some clarification on what to do.

```rust
#[doc=r#"without a newline"#]
#[doc=r#"with a
newline included"#]
```

would become i.e.

```rust
#[doc=r#"without a newline with a newline"#]
#[doc=r#"newline included "#]
```

**OR**


```rust
#[doc=r#"without a newline with a newline
                newline included "#]
```

And that's a valid case we should definitely decide, either way, we just should define behviour explicitly and write test cases accordingly.

drahnr commented on src/action/bandaidset.rs:
```diff
@@ -231,8 +231,8 @@ pub(crate) mod tests {
-                span: (2_usize, 3..43).try_into().unwrap(),
-                replacement: " automate that. Maybe not. But writing this is annoying".to_owned(),
+                span: (2_usize, 0..43).try_into().unwrap(),
+                replacement: "/// automate that. Maybe not. But writing this is annoying".to_owned(),
```
but why would we replace the `///` ? we want to preserve them, no? That way we become invariant to the comment type.

KuabeM replied:
At some point, we do have to add `///` for additional lines. In my opinion, it was more concise to add `/// ` on all multiline replacements rather than checking if it is a newly added line. 

However, if we took the approach of splitting `BandAid` into an enum, the above approach will be better. As this option is also a good choice for dealing with deletion/addition of lines in general, I suggest I'll start implementing it.

drahnr replied:
> At some point, we do have to add `///` for additional lines. In my opinion, it was more concise to add `/// ` on all multiline replacements rather than checking if it is a newly added line.

My point is rather, that we don't necessarily. We only need to add additional lines for `///`, but not necessarily for `#[doc=".."]` (depending on the discussion above).

> However, if we took the approach of splitting `BandAid` into an enum, the above approach will be better. As this option is also a good choice for dealing with deletion/addition of lines in general, I suggest I'll start implementing it.

:+1:

KuabeM replied:
In any case, I think its much simpler to refactor `BandAid` :)

drahnr commented on src/action/bandaidset.rs:
```diff
@@ -253,12 +253,12 @@ pub(crate) mod tests {
-                span: (2_usize, 3..61).try_into().unwrap(),
-                replacement: " automate that. Maybe not. But writing this is annoying.".to_owned(),
+                span: (2_usize, 0..61).try_into().unwrap(),
+                replacement: "/// automate that. Maybe not. But writing this is annoying.".to_owned(),
```
CC

drahnr commented on src/action/bandaidset.rs:
```diff
@@ -253,12 +253,12 @@ pub(crate) mod tests {
-                span: (3_usize, 3..37).try_into().unwrap(),
-                replacement: " However, I don't have a choice now, do I? Come on!".to_owned(),
+                span: (3_usize, 0..37).try_into().unwrap(),
+                replacement: "/// However, I don't have a choice now, do I? Come on!".to_owned(),
```
CC

KuabeM commented:
First stop, I implemented the refactoring with a `BandAid` enum. I reached a state where everything works as before it. I will now tackle the remaining issues:

1. Fix tests for `BandAid`/add more tests for `correct_lines()`
2. Don't include `///` `//!` in the replacement string of a `BandAid`, rather write it on `Injections`
3. Implement a clear and concise solution for newlines in `#[ doc =` comments.

Did I miss anything?

drahnr commented on src/action/bandaidset.rs:
```diff
@@ -133,7 +133,7 @@ impl FirstAidKit {
                     // Original content is shorter than replacement
                     let insertion = LineColumn {
-                        line: span.end.line,
+                        line: span.end.line + 1,
```
No `+1` without an explicit explanation why this is needed. That cause a shit ton of refactoring before :) and I don't feel a need to repeat that mistake.

KuabeM replied:
Yeah I see that :smile: I try to avoid it in the first place, but I don't think its possible :thinking:

drahnr commented on src/action/bandaidset.rs:
```diff
@@ -65,7 +64,7 @@ impl FirstAidKit {
         );
 
-        if span.is_multiline() {
+        if span.is_multiline() || replacement.lines().count() > 1 {
```
So you are checking if either is multiline, but what's the purpose? I would expect that we do not need this check, but just pick one of the iterators and consume it entirely, and drive the second iterator manually, any excess elements of the second are handled as `Insertion` / `Deletion` (depending on the definition of your first and second iterator).

Does that make sense?

drahnr replied:
I think you already impl'd it like that, but the `if` clause around the first part is kind of pointless.

KuabeM replied:
Yes should not matter

drahnr replied:
Then lets put it to rest :wastebasket: :smiley: 

less code, less :bug:s :)

KuabeM replied:
Actually, that may avoid some performance overhead since most replacements and spans will only be single-lined. That means we don't need to compute all that stuff inside this if clause.

drahnr commented on src/action/bandaidset.rs:
```diff
@@ -104,15 +103,11 @@ impl FirstAidKit {
+                    // Replacement covers a line in original content
+
+                    // get the length of the current line in the original content
                     let end_of_line: Vec<usize> = chunk
```
This would inaccurate for i.e.:

```rust
#[doc = r#"0"#]
#[doc = r#"a
c
b"#]
#[doc = r#"inf"#]
```

KuabeM replied:
That thing is entirely wrong I know. I'm on my way implementing a proper solution.

drahnr commented:
> First stop, I implemented the refactoring with a `BandAid` enum. I reached a state where everything works as before it. I will now tackle the remaining issues:
> 
>     1. Fix tests for `BandAid`/add more tests for `correct_lines()`
>     2. Don't include `///` `//!` in the replacement string of a `BandAid`, rather write it on `Injections`
>     3. Implement a clear and concise solution for newlines in `#[ doc =` comments.
> 
> 
> Did I miss anything?

Looks good! I can't think of anything besides that.

drahnr commented:
Another rebase round to get CI going.

drahnr commented:
Progress! :steam_locomotive: 

There is still one assertion that fails, but you are getting closer!

```sh
> cargo run -- reflow --code 27 -r demo/src/

    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
     Running `target/debug/cargo-spellcheck reflow --code 27 -r demo/src/`
[2020-10-07T07:15:31Z ERROR cargo_spellcheck::documentation::literalset] Adjacent literal is not the same comment variant: TripleSlash vs DoubleSlashEM
[2020-10-07T07:15:31Z ERROR cargo_spellcheck::documentation::literalset] Adjacent literal is not the same comment variant: MacroDocEq vs TripleSlash
[2020-10-07T07:15:31Z ERROR cargo_spellcheck::documentation::literalset] Adjacent literal is not the same comment variant: TripleSlash vs MacroDocEq
[2020-10-07T07:15:31Z ERROR cargo_spellcheck::documentation::literalset] Adjacent literal is not the same comment variant: MacroDocEq vs TripleSlash
[2020-10-07T07:15:31Z ERROR cargo_spellcheck::documentation::literalset] Adjacent literal is not the same comment variant: TripleSlash vs MacroDocEq
[2020-10-07T07:15:31Z ERROR cargo_spellcheck::documentation::literalset] Adjacent literal is not the same comment variant: TripleSlash vs DoubleSlashEM
[2020-10-07T07:15:31Z ERROR cargo_spellcheck::documentation::literalset] Adjacent literal is not the same comment variant: TripleSlash vs DoubleSlashEM
[2020-10-07T07:15:31Z ERROR cargo_spellcheck::documentation::literalset] Adjacent literal is not the same comment variant: MacroDocEq vs TripleSlash
[2020-10-07T07:15:31Z ERROR cargo_spellcheck::documentation::literalset] Adjacent literal is not the same comment variant: TripleSlash vs MacroDocEq
error: spellcheck(Reflow)
   --> /media/supersonic1t/projects/cargo-spellcheck/demo/src/nested/mod.rs:11
    |
 11 |  Overly long statements that should be reflown since they are __very__ long and exceed the line limit.
    |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

(1/6) Apply this suggestion [y,n,q,a,d,j,e,?]?

 »  Overly long statements that should be reflown since they are __very__ long
   ...                                                                        /// and exceed the line limit.
thread 'main' panicked at 'Extracting `Bandaid`s from `State` must not fail. qed: BUG: Found a range which does not cover its own chunk', src/action/interactive.rs:162:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

and one test is still going :boom: 

```
thread 'documentation::chunk::test::find_line_length_docmacro' panicked at 'assertion failed: `(left == right)`
  left: `20`,
 right: `23`', src/documentation/chunk.rs:786:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    documentation::chunk::test::find_line_length_docmacro
```

drahnr reviewed (commented):
Nice! :+1: I have yet to test it, but looks good!

drahnr commented on src/action/bandaidset.rs:
```diff
@@ -47,15 +47,36 @@ impl TryFrom<(String, &Span)> for FirstAidKit {
         if span.is_multiline() {
             bail!("Can't construct `FirstAidKit` from single-line span only")
```
`bail!` text is in conflict with condition check

KuabeM replied:
good catch, thanks!

drahnr commented on src/documentation/literal.rs:
```diff
@@ -14,8 +14,8 @@ pub enum CommentVariant {
     DoubleSlashEM,
-    /// `#[doc=`
-    MacroDocEq,
+    /// `#[doc=` with the length of possible `r#`s
```
Is this the length of `r###` with or without `r` or just the number `#`?

KuabeM replied:
it's both, added a note to the comment

drahnr commented on src/documentation/literal.rs:
```diff
@@ -31,9 +31,17 @@ impl Default for CommentVariant {
+            CommentVariant::DoubleSlashEM => "//!".into(),
+            CommentVariant::MacroDocEq(p) => {
+                let raw = match p {
+                    0 => "".to_string(),
```
I am generally a bit suspicious when using `.to_string()` but as a matter of fact to it calls `.to_owned()` internally https://doc.rust-lang.org/src/alloc/str.rs.html#204-206 https://doc.rust-lang.org/src/alloc/string.rs.html#2272-2277

So all my comments before were kind of pointless here, except for saving a few chars :man_shrugging:

KuabeM replied:
what do you suggest?  Try to drop all `to_string()` and use `&str` instead?

drahnr replied:
Mostly informational :) code is fine as is

drahnr reviewed (commented):
With this change you cannot represent insertions with BandAid anymore, can you? A span is by definition at least length 1 since the start and end members are inclusive.

KuabeM commented:
> With this change you cannot represent insertions with BandAid anymore, can you? A span is by definition at least length 1 since the start and end members are inclusive.

We don't need any insertions. Any BandAid always replaces something.
